In [5]:
import csv

ages = []
sexes = []
bmis = []
children = []
smoker = []
region = []
bmi_numbers = []
charges = []

def create_list_data(converted_list, insurance_csv, column_name):
    with open('insurance.csv', 'r+') as insurance_csv:
        csv_dict = csv.DictReader(insurance_csv)
        for row in csv_dict:
            converted_list.append(row[column_name])
        return converted_list

create_list_data(ages, 'insurance.csv', 'age')
create_list_data(sexes, 'insurance.csv', 'sex')
create_list_data(bmis, 'insurance.csv', 'bmi')
create_list_data(children, 'insurance.csv', 'children')
create_list_data(smoker, 'insurance.csv', 'smoker')
create_list_data(region, 'insurance.csv', 'region')
create_list_data(charges, 'insurance.csv', 'charges')

#for those which are not integers converted into floating numbers.
bmi_numbers = [float(i) for i in bmis]
children_number = [int(i) for i in children]
charges_number = [float(i) for i in charges]
ages_number = [float(i) for i in ages]


We have created lists where all the related data is held to related person.

In [6]:
class Patients:
    def __init__(self, patients_ages, patients_sexes, patients_bmis, patients_num_children, 
                 patients_smoker_statuses, patients_regions, patients_charges):
        self.patients_ages = patients_ages
        self.patients_sexes = patients_sexes
        self.patients_bmis = patients_bmis
        self.patients_num_children = patients_num_children
        self.patients_smoker_statuses = patients_smoker_statuses
        self.patients_regions = patients_regions
        self.patients_charges = patients_charges

    def analyze_ages(self):
        sum_age = 0 
        for age in self.patients_ages:
            sum_age += age
        return ("The average age of the patients is :" + str(sum_age / len(self.patients_ages)))
    
    def min_max_ages(self):
        min_age = min(self.patients_ages)
        max_age = max(self.patients_ages)
        print("The youngest person's age is: " + str(min_age) + '\n' + "The oldest person's age is: " + str(max_age))

    def analyze_sexes(self):
        females=0
        males=0
        for sex in self.patients_sexes:
            if sex == "male":
                males +=1
            elif sex == "female":
                females +=1
        print("The " + str(round((males/len(self.patients_ages)*100),2)) + "% of the population is male, and " + str(round(100-((males/len(self.patients_ages)*100)),2)) + "% of the poulation is female")
        if abs(round((males/len(self.patients_ages)*100),2) - 50) <= 4:
            print("The data is almost evenly distributed in sex-wise")

    def analyze_bmis(self):
        total_bmi = 0
        for bmi in self.patients_bmis:
            total_bmi += bmi
        print("Average bmi index is: " + str(round(total_bmi/ len(self.patients_bmis),2)))


    def unique_region(self):
        regions = {}
        for region in self.patients_regions:
            regions[region] = self.patients_regions.count(region)
        print(regions)

    def analyze_smoker_ratio(self):
        smoker_total = 0
        non_smoker_total = 0
        no_info = 0
        for status in self.patients_smoker_statuses:
            if status == "yes":
                smoker_total+=1
            elif status == "no":
                non_smoker_total+=1
            else:
                no_info+=1
        smoker_ratio = round((smoker_total/len(self.patients_smoker_statuses))*100,2)
        nonsmoker_ratio =round((non_smoker_total/len(self.patients_smoker_statuses))*100,2)
        print(str(smoker_ratio) + "% of the population smokes whereas " + str(nonsmoker_ratio) + "% are non-smoker.")



We have created a class named Patient which we conducted analysis that will be explained below. Before that we will create patients_info to execute the analysis for the data

In [7]:
patients_info = Patients(ages_number, sexes, bmi_numbers, children_number, smoker, region, charges_number)

In [8]:
patients_info.analyze_bmis()

Average bmi index is: 30.66


"BMI is a person’s weight in kilograms divided by the square of height in meters." according to the CDC. There are for ranges defined on for bmi. The average bmi index falls in the obese range.

In [9]:
patients_info.analyze_sexes()

The 50.52% of the population is male, and 49.48% of the poulation is female
The data is almost evenly distributed in sex-wise


In [28]:
regions = patients_info.unique_region()

{'southwest': 325, 'southeast': 364, 'northwest': 325, 'northeast': 324}


The data's distribution is fair region-wise.

In [11]:
region_base_bmi = {}
i = 0
while i < len(bmi_numbers):
    if region[i] not in region_base_bmi:
        region_base_bmi[region[i]] = [bmi_numbers[i]]
    elif region[i] in region_base_bmi:
        region_base_bmi[region[i]].append(bmi_numbers[i])
    i+=1
 
#Average BMI indexes of regions
southwest_avg = (round(float(sum(d for d in region_base_bmi["southwest"])) / len(region_base_bmi["southwest"]),2))
southeast_avg = (round(float(sum(d for d in region_base_bmi["southeast"])) / len(region_base_bmi["southeast"]),2))
northwest_avg = (round(float(sum(d for d in region_base_bmi["northwest"])) / len(region_base_bmi["northwest"]),2))
northeast_avg = (round(float(sum(d for d in region_base_bmi["northeast"])) / len(region_base_bmi["northeast"]),2))

print("Southwest region's average BMI: " + str(southwest_avg))
print("Southeast region's average BMI: " + str(southeast_avg))
print("Northwest region's average BMI: " + str(northwest_avg))
print("Northeast region's average BMI : " + str(northeast_avg))

Southwest region's average BMI: 30.6
Southeast region's average BMI: 33.36
Northwest region's average BMI: 29.2
Northeast region's average BMI : 29.17


Northeast and Northwest regions are in overweighted range, Southwest and Southeast regions are in obese range.
So we should separete regions while calculating charges based on BMI to see the bmi effect. 

But first, let's see if other parameters have pretty same effects in each and every region.
We will check if smokers are fairly distributed, the regions average age, average number of children people have and gender-based distribution per region.

In [12]:
region_based_smokers_ratio = {}
i = 0
while i < len(bmi_numbers):
    if region[i] not in region_based_smokers_ratio:
        region_based_smokers_ratio[region[i]] = [smoker[i]]
    elif region[i] in region_based_smokers_ratio:
        region_based_smokers_ratio[region[i]].append(smoker[i])
    i+=1

southwest_smokers_ratio = round((region_based_smokers_ratio["southwest"].count('yes')*100)/len(region_based_smokers_ratio["southwest"]),2)
southeast_smokers_ratio = round((region_based_smokers_ratio["southeast"].count('yes')*100)/len(region_based_smokers_ratio["southeast"]),2)
northwest_smokers_ratio = round((region_based_smokers_ratio["northwest"].count('yes')*100)/len(region_based_smokers_ratio["northwest"]),2)
northeast_smokers_ratio = round((region_based_smokers_ratio["northeast"].count('yes')*100)/len(region_based_smokers_ratio["northeast"]),2)

print("southwest_smokers_ratio: " + str(southwest_smokers_ratio) + "%")
print("southeast_smokers_ratio: " + str(southeast_smokers_ratio) + "%")
print("northwest_smokers_ratio: " + str(northwest_smokers_ratio) + "%")
print("northeast_smokers_ratio: " + str(northeast_smokers_ratio) + "%")



southwest_smokers_ratio: 17.85%
southeast_smokers_ratio: 25.0%
northwest_smokers_ratio: 17.85%
northeast_smokers_ratio: 20.68%


Southeast's and Northeast's smoker population ratios are pretty high. So, it will be excluded for calculating bmi effect.

In [13]:
region_based_children = {}
region_based_children['southwest'] = []
region_based_children['northwest'] = []
region_based_children['southeast'] = []
region_based_children['northeast'] = []

i=0
while i < len(children_number):
    region_based_children[region[i]].append(children_number[i])
    i+=1

print("The average number of children per person for people in southwest is " + str(round(sum(d for d in region_based_children['southwest'])/len(region_based_children['southwest']),2)))
print("The average number of children per person for people in northwest is " + str(round(sum(d for d in region_based_children['northwest'])/len(region_based_children['northwest']),2)))



The average number of children per person for people in southwest is 1.14
The average number of children per person for people in northwest is 1.15


Number of children people have on average are pretty same for every region.
We will continue with Southwest and Northwest regions.

In [14]:
region_based_age = {}
region_based_age['southwest'] = []
region_based_age['northwest'] = []
region_based_age['southeast'] = []
region_based_age['northeast'] = []

i=0
while i < len(children_number):
    region_based_age[region[i]].append(ages_number[i])
    i+=1
    
print("The average age of people in southwest is " + str(round(sum(d for d in region_based_age['southwest'])/len(region_based_age['southwest']),2)))
print("The average age of people in northwest is " + str(round(sum(d for d in region_based_age['northwest'])/len(region_based_children['northwest']),2)))
print("The average age of people in northeast is " + str(round(sum(d for d in region_based_age['northeast'])/len(region_based_age['northeast']),2)))
print("The average age of people in southeast is " + str(round(sum(d for d in region_based_age['southeast'])/len(region_based_children['southeast']),2)))


The average age of people in southwest is 39.46
The average age of people in northwest is 39.2
The average age of people in northeast is 39.27
The average age of people in southeast is 38.94


The average ages are consistant for Southwest and Northwest regions.

Next, we will check if the population is fair gender-wise per region

In [15]:
region_based_sex = {}
region_based_sex['southwest'] = []
region_based_sex['northwest'] = []
region_based_sex['southeast'] = []
region_based_sex['northeast'] = []

i=0

while i< len(children_number):
    region_based_sex[region[i]].append(sexes[i])
    i+=1
print("southwest female count: " + str(region_based_sex['southwest'].count('female')))
print("nortwest female count: " + str(region_based_sex['northwest'].count('female')))
print("southeast female count: " + str(region_based_sex['southeast'].count('female')))
print("northeast female count: "  + str(region_based_sex['northeast'].count('female')))

southwest female count: 162
nortwest female count: 164
southeast female count: 175
northeast female count: 161


Since their population is the same size, we can conclude that the distribution of the data is fair gender-wise.
Hence, we have two sample data being almost the same in the light of the the parameters we check.
Finally, we will see if the charges differ in bmi-wise.

In [16]:
region_base_charge = {}
i = 0
while i < len(bmi_numbers):
    if region[i] not in region_base_charge:
        region_base_charge[region[i]] = [charges_number[i]]
    elif region[i] in region_base_charge:
        region_base_charge[region[i]].append(charges_number[i])
    i+=1

southwest_avg_charge = (round(float(sum(d for d in region_base_charge["southwest"])) / len(region_base_charge["southwest"]),2))
southeast_avg_charge = (round(float(sum(d for d in region_base_charge["southeast"])) / len(region_base_charge["southeast"]),2))
northwest_avg_charge = (round(float(sum(d for d in region_base_charge["northwest"])) / len(region_base_charge["northwest"]),2))
northeast_avg_charge = (round(float(sum(d for d in region_base_charge["northeast"])) / len(region_base_charge["northeast"]),2))

print("Southwest region's average charge: " + str(southwest_avg_charge))
print("Northwest region's average charge: " + str(northwest_avg_charge))


Southwest region's average charge: 12346.94
Northwest region's average charge: 12417.58


Since average bmi is higher in Southwest, we would have expected the charges for Southwest to be higher. However, we can see that Southwest charges are lower than Nortwest. 
Based on the findings, we can only tell that there is no significant effect that lowering BMI decreases insurance charges.

We pretty much calculated every parameter for every region. Let's check if smoking ratio affects charges.
Again, we will continue analyzing parameters below. I have picked Northeast and Northwest regions because they look similar in terms of parameters except smoking ratio

In [17]:
print("northwest_smokers_ratio: " + str(northwest_smokers_ratio) + "%")
print("northeast_smokers_ratio: " + str(northeast_smokers_ratio) + "%")

northwest_smokers_ratio: 17.85%
northeast_smokers_ratio: 20.68%


In [18]:
print("The average number of children per person for people in northwest is " + str(round(sum(d for d in region_based_children['northwest'])/len(region_based_children['northwest']),2)))
print("The average number of children per person for people in northeast is " + str(round(sum(d for d in region_based_children['northeast'])/len(region_based_children['northeast']),2)))


The average number of children per person for people in northwest is 1.15
The average number of children per person for people in northeast is 1.05


People have on average pretty same number of children in Northwest and Northeast.

In [19]:
print("The average age of people in northwest is " + str(round(sum(d for d in region_based_age['northwest'])/len(region_based_children['northwest']),2)))
print("The average age of people in northeast is " + str(round(sum(d for d in region_based_age['northeast'])/len(region_based_age['northeast']),2)))

The average age of people in northwest is 39.2
The average age of people in northeast is 39.27


They are quite similar age-wise.

In [20]:
print("Northwest region's average BMI: " + str(northwest_avg))
print("Northeast region's average BMI : " + str(northeast_avg))

Northwest region's average BMI: 29.2
Northeast region's average BMI : 29.17


Bmi indexes are the same, we have checked just to be sure.

In [21]:
print("nortwest female count: " + str(region_based_sex['northwest'].count('female')))
print("northeast female count: "  + str(region_based_sex['northeast'].count('female')))


nortwest female count: 164
northeast female count: 161


The data is almost evenly distributed in gender-wise

In [22]:
print("Northwest region's average charge: " + str(northwest_avg_charge))
print("Northeast region's average charge: " + str(northeast_avg_charge))
difference = round(((northeast_avg_charge-northwest_avg_charge)*100)/northeast_avg_charge,2)
print("The difference is:" + str(difference) +"%")

Northwest region's average charge: 12417.58
Northeast region's average charge: 13406.38
The difference is:7.38%


Let's remember the smoking ratios:

northwest_smokers_ratio: 17.85%
northeast_smokers_ratio: 20.68%

Based on the findings, we can definitely say 3% increase in smoking ratio(northeast_smokers_ratio-northwest_smokers_ratio) increases charges by 7%.

With the help of math:
if 3% increases charges by 7% on average: 20.68% decrease of smoker-ratio in Northeast should be expected to lower charges by 62.04% (20.68*3) on average, approximately.

In [27]:
region_based_charges_smoking = {
    'southeast': {'yes': [], 'no': []},
    'northeast': {'yes': [], 'no': []},
    'northwest': {'yes': [], 'no': []},
    'southwest': {'yes': [], 'no': []},
}

i=0
for place in region:
    if smoker[i] == 'yes':
            region_based_charges_smoking[region[i]][smoker[i]].append(charges_number[i])
    elif smoker[i] == 'no':
            region_based_charges_smoking[region[i]][smoker[i]].append(charges_number[i])
    i+=1

average_smoker_southeast = round(sum(d for d in region_based_charges_smoking['southeast']['yes'])/len(region_based_charges_smoking['southeast']['yes']),2)
average_nonsmoker_southeast = round(sum(d for d in region_based_charges_smoking['southeast']['no'])/len(region_based_charges_smoking['southeast']['no']),2)
average_smoker_southwest = round(sum(d for d in region_based_charges_smoking['southwest']['yes'])/len(region_based_charges_smoking['southwest']['yes']),2)
average_nonsmoker_southwest = round(sum(d for d in region_based_charges_smoking['southwest']['no'])/len(region_based_charges_smoking['southwest']['no']),2)

average_smoker_northeast = round(sum(d for d in region_based_charges_smoking['northeast']['yes'])/len(region_based_charges_smoking['northeast']['yes']),2)
average_nonsmoker_northeast = round(sum(d for d in region_based_charges_smoking['northeast']['no'])/len(region_based_charges_smoking['northeast']['no']),2)
average_smoker_northwest = round(sum(d for d in region_based_charges_smoking['northwest']['yes'])/len(region_based_charges_smoking['northwest']['yes']),2)
average_nonsmoker_northwest = round(sum(d for d in region_based_charges_smoking['northwest']['no'])/len(region_based_charges_smoking['northwest']['no']),2)

print(
    "Smokers pay " + str(round((((average_smoker_northeast-average_nonsmoker_northeast)*100)/average_smoker_northeast),2)) + "% more than non-smokers in Northeast"
)

print(
    "Smokers pay " + str(round((((average_smoker_northwest-average_nonsmoker_northwest)*100)/average_smoker_northeast),2)) + "% more than non-smokers in Northwest"
)



Smokers pay 69.11% more than non-smokers in Northeast
Smokers pay 72.91% more than non-smokers in Northwest


In latest calculation we should conduct more researches on given groups for the parameters. However this finding supports that  quitting smoking lowers the charges at least 62.04% in Northeast. 